pkg/simple/client/monitoring/metricsserver/metricsserver.go

```diff
@@ -0,0 +1,368 @@
+/*
+Copyright 2020 KubeSphere Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricsserver
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+
+	metricsapi "k8s.io/metrics/pkg/apis/metrics"
+	metricsV1beta1 "k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	metricsclient "k8s.io/metrics/pkg/client/clientset/versioned"
+
+	"kubesphere.io/kubesphere/pkg/simple/client/k8s"
+	"kubesphere.io/kubesphere/pkg/simple/client/monitoring"
+)
+
+// metricsServer implements monitoring interface backend by metrics-server
+type metricsServer struct {
+	metricsAPIAvailable bool
+	metricsClient       *metricsclient.Clientset
+	k8s                 kubernetes.Interface
+}
+
+var (
+	supportedMetricsAPIs = map[string]bool{
+		"v1beta1": true,
+	}
+)
+
+const edgeNodeLabel = "node-role.kubernetes.io/edge"
+
+func metricsAPISupported(discoveredAPIGroups *metav1.APIGroupList) bool {
+	for _, discoveredAPIGroup := range discoveredAPIGroups.Groups {
+		if discoveredAPIGroup.Name != metricsapi.GroupName {
+			continue
+		}
+		for _, version := range discoveredAPIGroup.Versions {
+			if _, found := supportedMetricsAPIs[version.Version]; found {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func intersect(a map[string]bool, b []string) map[string]bool {
+	set := make(map[string]bool)
+
+	for _, el := range b {
+		if _, found := a[el]; found {
+			set[el] = true
+		}
+	}
+
+	return set
+}
+
+func (m metricsServer) listEdgeNodes() (map[string]v1.Node, error) {
+	nodes := make(map[string]v1.Node)
+
+	nodeClient := m.k8s.CoreV1()
+
+	nodeList, err := nodeClient.Nodes().List(context.TODO(), metav1.ListOptions{
+		LabelSelector: edgeNodeLabel,
+	})
+	if err != nil {
+		return nodes, err
+	}
+
+	for _, n := range nodeList.Items {
+		nodes[n.Name] = n
+	}
+
+	return nodes, nil
+}
+
+func (m metricsServer) filterEdgeNodeNames(edgeNodes map[string]v1.Node, opts *monitoring.QueryOptions) map[string]bool {
+	edgeNodeNamesMap := make(map[string]bool)
+
+	for _, n := range edgeNodes {
+		edgeNodeNamesMap[n.Name] = true
+	}
+
+	if opts.ResourceFilter == ".*" {
+		return edgeNodeNamesMap
+	}
+
+	resourceFilter := strings.TrimSuffix(opts.ResourceFilter, "$")
+	nodeNames := strings.Split(resourceFilter, "|")
+
```

Comment:
The resource filter parameter described by the API may be a regexp pattern, although the Console only uses it as a string of resource names concatenated by `|`

Reply:
Can you provide detail specification on this filter parameter?

Reply:
although it's a regex, but basically we only use it in form like `node1|node2|node3`, right? @junotx

Reply:
Yes, please click [here](https://github.com/kubesphere/kubesphere/blob/release-3.0/api/ks-openapi-spec/swagger.json#L6822) for the description of this filter parameter

Reply:
We need to handle inputs not following concating by | form.

Reply:
Fixed with prometheus regex matcher.